### ISSUE TITLE
Bug: SimCLRLoss has a mistake in normalizing za and zb

### ISSUE THREAD

**Issue by kechan** [tensorflow/similarity#237]
**Where:** 

https://github.com/tensorflow/similarity/blob/master/tensorflow_similarity/losses/simclr.py

In def call(...):

```
if self.use_hidden_norm:
    za = tf.math.l2_normalize(za)
    zb = tf.math.l2_normalize(zb)
```

Should be:

```
if self.use_hidden_norm:
    za = tf.math.l2_normalize(za, axis=-1)
    zb = tf.math.l2_normalize(zb, axis=-1)
```

**Cause**
Doc: https://www.tensorflow.org/api_docs/python/tf/math/l2_normalize

Default axis is None, so the normalization will be over batch as well as embedding dim. We want only norm for embedding vec (ie. axis=-1).

**Consequence**

SimCLR training failed with loss stuck at an unreasonably high level even with 1 batch of data and ResNet50Sim.

**Related**
See Issue https://github.com/tensorflow/similarity/issues/233 on full detail on the test and investigation behind this.

**Comment by owenvallis** [comment]
Looks like the same issue in the barlow twins loss. I've changed them both and I'm rerunning the notebook. I'll push the patch once I verify the results are better, but so far it's looking good for SimCLR. Thanks again for finding this.

**Comment by kechan** [comment]
Glad to hear this helps with debugging the barlow. 

Just throwing out another thought unrelated to this issue. 
Is the dataset shuffle buffer size of 1024 coming from their papers or code. I would be surprised in 3 papers used the same dataset setup. Since the batch size is 512, i thought the shuffle buffer size could be a bit bigger. 

I don't have deep enough experience with contrastive training. So this is speculative. I have a hunch that higher quality shuffling may lead to more optimal training that produce better representation. Not sure if there's any research done that put some emphasis on this (or it doesn't matter). I may probably play around with this and see.
